classes/helper.php

```diff
@@ -86,29 +87,40 @@ public static function build_rules_from_config() {
         return $rules;
     }
 
-    /**
-     * Build rules from DB.
-     *
-     * @throws \coding_exception
-     */
-    public static function build_rules_from_db() {
-        throw new \coding_exception('Not implemented yet! Please use helper::build_rules_from_config');
-    }
-
     /**
      * Redirects based on rules
      */
     public static function redirect_from_rules() {
-        global $PAGE;
+        global $FULLME;
 
-        if ($PAGE->has_set_url()) {
-            $rules = self::get_all_rules();
+        static $processed = false;
 
-            foreach ($rules as $rule) {
-                if ($rule->is_enabled() && $rule->should_redirect($PAGE->url)) {
-                    redirect($rule->get_redirect_url());
+        if ($processed) {
+            return;
+        }
+
+        $rules = self::get_all_rules();
+
+        foreach ($rules as $rule) {
+            if ($rule->is_enabled() && $rule->should_redirect(new \moodle_url($FULLME))) {
+
+                $target = $rule->get_redirect_url();
+
+                // Check of backdoor for admins in case of a horrible mistake happened.
+                if ($rule->should_warn_instead_of_redirect()) {
+                    \core\notification::info(get_string('redirectwarning', 'tool_redirects', [
+                        'target'  => $target->out(),
+                        'regex'   => $rule->get_regex(),
+                        'editurl' => (new \moodle_url('/admin/settings.php?section=tool_redirects'))->out(),
+                    ]));
+                    break;
                 }
+
+                redirect($target);
             }
         }
+
+        // Never process more than once, because we listen to more than one callback.
+        $processed = true;
```

Comment:
Is this param meant to stop infinite loops ?  Or just purely for limiting only 1 rule being processed? 

I was able to make it infinitely loop by a rule: `#\/my#=>/my`

Reply:
Infinite loops are probably not such a big issue since it would be just bad config from the admin side and would be pretty easy to spot. Just was curious if this was intended to stop them (and isn't working) or if its for something else entirely.

Reply:
No this is because of listening to _before_http_headers and _after_config and when in debug mode I only want to show one notification

Redirect loops could be multi step and are very tricky to protect from, don't want to attempt it here

Reply:
Yep that makes sense. :+1: